File: cpp/06_edge/main.cpp
// 06_edge: Sobel / Laplacian / Canny によるエッジ検出
// Usage: ./06_edge <input_image>
#include <opencv2/opencv.hpp>
#include "common/utils.hpp"

int main(int argc, char** argv) {
    const std::string in_path = tutorial::arg_or(argc, argv, 1, "../data/building.jpg");
    cv::Mat gray = tutorial::load_image(in_path, cv::IMREAD_GRAYSCALE);

    // ノイズ低減
    cv::Mat blur;
    cv::GaussianBlur(gray, blur, cv::Size(3, 3), 0);

    // 1) Sobel (x, y)
    cv::Mat gx, gy, abs_gx, abs_gy, sobel_mag;
    cv::Sobel(blur, gx, CV_16S, 1, 0, 3);
    cv::Sobel(blur, gy, CV_16S, 0, 1, 3);
    cv::convertScaleAbs(gx, abs_gx);
    cv::convertScaleAbs(gy, abs_gy);
    cv::addWeighted(abs_gx, 0.5, abs_gy, 0.5, 0, sobel_mag);

    // 2) Laplacian
    cv::Mat lap, abs_lap;
    cv::Laplacian(blur, lap, CV_16S, 3);
    cv::convertScaleAbs(lap, abs_lap);

    // 3) Canny
    cv::Mat canny;
    cv::Canny(blur, canny, 80, 160);

    tutorial::show_and_wait({
        {"gray", gray},
        {"sobel x", abs_gx},
        {"sobel y", abs_gy},
        {"sobel mag", sobel_mag},
        {"laplacian", abs_lap},
        {"canny 80/160", canny},
    });
    return 0;
}

File: python/06_edge/main.py
"""06_edge: エッジ検出 (Sobel / Laplacian / Canny / Scharr)
Usage (CLI): python main.py [input_image]
"""
import argparse
import sys
from pathlib import Path

import cv2
import numpy as np

sys.path.insert(0, str(Path(__file__).parent.parent))
from utils import DATA_DIR, load_image, show


def run(in_path: str | Path):
    src = load_image(in_path)
    gray = cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)

    # 前処理: ガウシアンブラーでノイズ軽減
    blurred = cv2.GaussianBlur(gray, (5, 5), 0)

    # 1) Sobel (X方向・Y方向・合成)
    sobelx = cv2.Sobel(blurred, cv2.CV_64F, 1, 0, ksize=3)
    sobely = cv2.Sobel(blurred, cv2.CV_64F, 0, 1, ksize=3)
    sobel_mag = cv2.magnitude(sobelx, sobely)
    sobel_mag = np.clip(sobel_mag, 0, 255).astype(np.uint8)
    sobelx_u8 = cv2.convertScaleAbs(sobelx)
    sobely_u8 = cv2.convertScaleAbs(sobely)

    # 2) Scharr (Sobel より精度高い)
    scharrx = cv2.Scharr(blurred, cv2.CV_64F, 1, 0)
    scharry = cv2.Scharr(blurred, cv2.CV_64F, 0, 1)
    scharr_mag = cv2.magnitude(scharrx, scharry)
    scharr_mag = np.clip(scharr_mag, 0, 255).astype(np.uint8)

    # 3) Laplacian
    lap = cv2.Laplacian(blurred, cv2.CV_64F, ksize=3)
    lap_u8 = cv2.convertScaleAbs(lap)

    # 4) Canny (ヒステリシス閾値)
    canny_lo = cv2.Canny(blurred, 50, 150)
    canny_hi = cv2.Canny(blurred, 100, 200)

    pairs = [
        ("gray", gray),
        ("Sobel X", sobelx_u8),
        ("Sobel Y", sobely_u8),
        ("Sobel magnitude", sobel_mag),
        ("Scharr magnitude", scharr_mag),
        ("Laplacian", lap_u8),
        ("Canny (50,150)", canny_lo),
        ("Canny (100,200)", canny_hi),
    ]
    return pairs


def main():
    parser = argparse.ArgumentParser(description="06_edge: エッジ検出")
    parser.add_argument("input", nargs="?", default=str(DATA_DIR / "lena.png"))
    parser.add_argument("--no-show", action="store_true")
    args = parser.parse_args()

    pairs = run(args.input)
    if not args.no_show:
        show(pairs, cols=4)


if __name__ == "__main__":
    main()
